common/src/types/v0/store/node.rs

```diff
@@ -66,33 +140,237 @@ impl NodeSpec {
+                        // remove the cordon label
+                        state.remove_cordon_label(label);
+                    }
+                },
+                None => {
+                    // should not be possible
+                }
+            }
+        } else if self.has_drain_label(label.clone()) {
+            // if the label is a drain label
+            match &mut self.cordon_drain_state {
+                Some(ds) => match ds {
+                    CordonDrainState::Cordoned(_) => {
+                        // should not be possible
+                    }
+                    CordonDrainState::Draining(state) => {
+                        // remove the drain label
+                        // if none left, set to cordoned or none
+                        state.remove_drain_label(label);
+                        if state.drainlabels.is_empty() {
+                            if state.cordonlabels.is_empty() {
+                                self.cordon_drain_state = None;
+                            } else {
+                                self.cordon_drain_state = Some(CordonDrainState::Cordoned(
+                                    CordonedState::new(state.cordonlabels.clone()),
+                                ));
+                            }
+                        }
+                    }
+                    CordonDrainState::Drained(state) => {
+                        // remove the drain label
+                        // if none left, set to cordoned or none
+                        state.remove_drain_label(label);
+                        if state.drainlabels.is_empty() {
+                            if state.cordonlabels.is_empty() {
+                                self.cordon_drain_state = None;
+                            } else {
+                                self.cordon_drain_state = Some(CordonDrainState::Cordoned(
+                                    CordonedState::new(state.cordonlabels.clone()),
+                                ));
+                            }
+                        }
+                    }
+                },
+                None => {
+                    // should not be possible
+                }
+            }
+        } else {
+            // return error
         }
     }
+
     /// Returns whether or not the node is cordoned.
     pub fn cordoned(&self) -> bool {
-        !self.cordon_labels.is_empty()
+        match &self.cordon_drain_state {
+            Some(ds) => match ds {
+                CordonDrainState::Cordoned(_) => true,
+                CordonDrainState::Draining(_) => true,
+                CordonDrainState::Drained(_) => true,
+            },
+            None => false,
+        }
+    }
+
+    /// Returns true if it has the label in either of the lists
+    pub fn has_cordon_label(&self, label: String) -> bool {
+        if self.has_cordon_only_label(label.clone()) || self.has_drain_label(label) {
```

Comment:
we could have jiust returned `self.has_cordon_only_label(label.clone()) || self.has_drain_label(label)` here instead of extra condition checking